lib/filesystem.ml

```diff
@@ -0,0 +1,240 @@
+    let d = Int64.unsigned_div ns ns_per_day
+    and ps = Int64.(mul (unsigned_rem ns ns_per_day) ps_per_ns)
+    in
+    Ptime.v (Int64.to_int d, ps)
+
+  let encode_timestamp data off v =
+    let d, ps = Ptime.Span.to_d_ps (Ptime.to_span v) in
+    let ns = Int64.(add (mul (Int64.of_int d) ns_per_day) (div ps ps_per_ns)) in
+    Bytes.set_int64_be data off ns
+
+  let safe_int ~msg d =
+    if d > Int64.of_int max_int then
+      Error (`Overflow msg)
+    else
+      Ok (Int64.to_int d)
+
+  type decode_err =
+    [ `Overflow of string |`Bad_checksum | `Bad_superblock_version of int ]
+
+  let pp_decode_err ppf = function
+    | `Overflow msg -> Fmt.pf ppf "integer overflow %s" msg
+    | `Bad_checksum -> Fmt.string ppf "bad superblock checksum"
+    | `Bad_superblock_version v ->
+      Fmt.pf ppf "superblock version %d is not supported (supported is %d)"
+        v s_version
+
+  let decode_superblock buf : (superblock, [> decode_err ]) result =
+    let payload, checksum =
+      let mid = String.length buf - H.digest_size in
+      String.sub buf 0 mid, String.sub buf mid H.digest_size
+    in
+    if String.equal checksum H.(to_raw_string (digest_string payload)) then
+      let super_version = String.get_uint16_be payload 0
+      and super_counter = String.get_int64_be payload 8
+      and timestamp = decode_timestamp payload 16
+      and active_sector = String.get_int64_be payload 24
+      and data_length = String.get_int64_be payload 32
+      and data_checksum = String.sub payload 40 H.digest_size
+      in
+      let ( let* ) = Result.bind in
+      if super_version = s_version then
+        let* super_counter = safe_int ~msg:"superblock counter" super_counter in
+        let* data_length = safe_int ~msg:"data length" data_length in
+        Ok { super_version ; super_counter ; timestamp ; active_sector ;
+             data_length ; data_checksum ; used_sectors = IS.empty }
+      else
+        Error (`Bad_superblock_version super_version)
+    else
+      Error `Bad_checksum
+
+  let encode_superblock t buf =
+    Bytes.set_uint16_be buf 0 t.super_version;
+    Bytes.set_int64_be buf 8 (Int64.of_int t.super_counter);
+    encode_timestamp buf 16 t.timestamp;
+    Bytes.set_int64_be buf 24 t.active_sector;
+    Bytes.set_int64_be buf 32 (Int64.of_int t.data_length);
+    Bytes.blit_string t.data_checksum 0 buf 40 H.digest_size;
+    let eop = Bytes.length buf - H.digest_size in
+    let payload = Bytes.sub_string buf 0 eop in
+    let checksum = H.(to_raw_string (digest_string payload)) in
+    Bytes.blit_string checksum 0 buf eop H.digest_size
+
+  let lwt_err_to_msg ~pp_error f =
+    let open Lwt.Infix in
+    f >|= Result.map_error (fun e -> `Msg (Fmt.to_to_string pp_error e))
+
+  let read_data block =
+    let open Lwt.Infix in
+    Block.get_info block >>= fun info ->
+    let open Lwt_result.Infix in
+    let ss = info.Mirage_block.sector_size in
+    assert (ss >= superblock_size);
```

Comment:
TODO: we should get rid of assertions in the code. log or return error...

Reply:
I saw this and added a check in `connect`. We can safely remove this `assert`.
```suggestion
```